### PR TITLE
Fixes #21987 - New UI for auth sources

### PR DESCRIPTION
**Pull request by dhlavac**
This PR changes Administration for Authentication sources

Now is by default added External auth source to database. In setting and for running systems by migration.
Every authentication source have card and display number of users set up by authentication sources

**Administer page for Authentication sources**
jhz1jwx

**Administer page for Authentication sources with no LDAPs**
auth_sources_empty

**Drop-downs for External and LDAP authentication sources**
auth_sources_with_external

### CODE REVIEW

**Review by amirfefer — COMMENTED** [review]
Thanks @dhlavac, looks great!
I'd like to see this card view as a react component.
there are patternfly-react [card](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Warning%20threshold%20value=40&knob-Header=Great%20job%21&knob-is%20Drawer%20Expandable=true&knob-Loading=true&knob-Description=Description&knob-animate=true&knob-Action=true&knob-Error%20threshold%20value=80&knob-Maximal%20value=100&knob-Message=This%20is%20really%20working%20out.&knob-Type=success&knob-Actual%20value=70&knob-Menu=true&knob-Size=md&knob-Match%20Height=true&knob-On%20Color=primary&knob-Off%20Color=default&knob-Minimal%20value=0&knob-Label=Label&selectedKind=patternfly-react%2FCards&selectedStory=Base%20Card&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel) and [kebab](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Warning%20threshold%20value=40&knob-Header=Great%20job%21&knob-is%20Drawer%20Expandable=true&knob-Loading=true&knob-Description=Description&knob-animate=true&knob-Action=true&knob-Error%20threshold%20value=80&knob-Maximal%20value=100&knob-Message=This%20is%20really%20working%20out.&knob-Type=success&knob-Actual%20value=70&knob-Menu=true&knob-Size=md&knob-Match%20Height=true&knob-On%20Color=primary&knob-Off%20Color=default&knob-Minimal%20value=0&knob-Label=Label&selectedKind=patternfly-react%2FDropdownKebab&selectedStory=DropdownKebab&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel) components already.

When a pull request is still in progress, you can add [WIP] on the PR title.

**Comment by ohadlevy** [comment]
I too would love to see it in react, but maybe prior to that we should review the UX with @rohoover as well? I could imagine different views (besides tables) that might be useful here?

**Comment by ares** [comment]
This is the outcome of discussion with @rohoover, it's a midterm solution, until we have org/loc association per external hostgroup, until them we need table for ldap auth sources, as users need to create many of them in multiorg scenario. After that's done, auth sources should be cards only.

**Comment by ohadlevy** [comment]
On Wed, May 16, 2018 at 8:59 AM, Marek Hulán <notifications@github.com>
wrote:

> This is the outcome of discussion with @Rohoover
> <https://github.com/Rohoover>, it's a midterm solution, until we have
> org/loc association per external hostgroup, until them we need table for
> ldap auth sources, as users need to create many of them in multiorg
> scenario. After that's done, auth sources should be cards only.
>
>
> Where can I find more information about your end goal?  I find it hard to
understand what you are trying to achieve in the long term? (sorry, but
just missing the user stories and dont understand how auth source, external
hostgroups and multiorg comes together)

**Comment by ares** [comment]
I don't recall if the meeting back then was recorded. Here's the motivation, today we can setup mapping between external auth sources and user groups, user groups can have roles assigned. With external user groups mapping, administrators can manage permissions externally to Foreman, only by assigning user groups in their LDAP. This works fine for permissions but not for ogranizations and locations, which is our secondary authorization system. For that we added orgs/locs associations to ldap auth sources. When user account is created, it gets associated to orgs/locs the LDAP is assigned to.

The problem is that if I want to have single LDAP for multiple organizations, each use belonging to different one, the model does not fit. People workaround that by having multiple LDAP definitions for the same LDAP server but with different filter and organization assigned. That means easily >10  LDAP authsources in env with multiple orgs.

We started adding the same model to external auth source (using just REMOTE_USER) but in order to do it, we had to add API and UI for manipulating it's org/loc association. As part of that, @dhlavac adds UI for it. This is a temporary solution, where we'll still have table for ldap authsources (as there can be a lot of them because of ^) and cards for internal and external (both have single instance). Later we want to add org/loc association also to external user group which should dramatically decrease the number of LDAP auth sources defined here and at that point we'd like to convert the table to cards. So at the end, all auth source instances should be represented by card.

Hope that puts more light on the change. When @dhlavac started working on this, there were no cards implemented and patternfly roadmap was unclear about this. Since it was few months ago, I think we can revisit the decision. Though I'm not sure I'd block this PR because of that single card partial. It's not the crucial part of it. But I'd definitely support the effort that would follow right after this gets in, so that we'll get card implemented as a react component if @dhlavac agrees.

**Comment by dhlavac** [comment]
@ares PR is ready I hope. @ohadlevy @amirfefer If it isn't problem, for now I will let card component in erb. It's a midterm solution, until we have org/loc association per external hostgroup and then implement react component for cards.

**Comment by ares** [comment]
There seems to be several hound/rubocop issues, mind to rebase and fix them?

**Review by ares — CHANGES_REQUESTED** [review]
rubocop failures, rebase would also be good (no conflicts)

File: config/environments/development.rb
Comment (ares) [inline config/environments/development.rb]: is this related?

Comment (dhlavac) [reply config/environments/development.rb]: Ou thanks ... leftover from debugging

**Comment by dhlavac** [comment]
@ares rebased and removed debug leftovers and hound violations

**Comment by ares** [comment]
this needs to be rebased

File: app/models/auth_source.rb
Comment (ares) [inline app/models/auth_source.rb]: would it be better to explicitly mention LDAP auth source class? if we add additional auth source in future, it will be also included in this only_ldap scope (in fact there's at least one plugin that adds it)

File: app/views/auth_source_ldaps/_form.html.erb
Comment (ares) [inline app/views/auth_source_ldaps/_form.html.erb]: missing space before %

File: db/migrate/20180720143228_set_default_authsource_external_setting.rb
Comment (ares) [inline db/migrate/20180720143228_set_default_authsource_external_setting.rb]: no need to extract strings here using N_(), they are already extracted in setting file, please remove from the whole migration

File: app/views/auth_sources/_card.html.erb
Comment (ares) [inline app/views/auth_sources/_card.html.erb]: it would be better to restructure this partial as it seems it's combining too many things, I'd suggest

1) make a generic card partial that contains the first three wrapper divs, it would accept variable body, it should also live in some more generic dir as it wouldn't be specific for auth sources
2) define auth_source_ldap_card and auth_source_external_card that would use the generic partial specifying their content as body, to avoid big if conditions in view layer
3) to share the common part I'd add one extra partial, auth_source_card that would lines 24-32 to the generic one and in fact partials from 2) would be using this auth_source_card instead of generic one directly

I hope that makes sense

File: app/views/auth_sources/index.html.erb
Comment (ares) [inline app/views/auth_sources/index.html.erb:9]: could you please explain this a bit more? the ldap_present method searches for first ldap auth source in `@auth_sources` if there's none it builds new one if my understanding is correct

also you already have ldap auth sources in `@auth_source_ldaps`, why filtering `@auth_sources`?

File: app/views/auth_sources/index.html.erb
Comment (ares) [inline app/views/auth_sources/index.html.erb:5]: you added scopes to AuthSource class, why not using them here?

**Review by ares — CHANGES_REQUESTED** [review]
couple of new findings, please take a look

File: app/views/auth_sources/_card.html.erb
Comment (dhlavac) [reply app/views/auth_sources/_card.html.erb]: Thanks, good idea. What do you think where can I place this generic card ?

File: app/views/auth_sources/index.html.erb
Comment (dhlavac) [reply app/views/auth_sources/index.html.erb:5]: We have scopes for only_ldaps but not for everything except ldaps(and hidden)

File: app/views/auth_sources/index.html.erb
Comment (dhlavac) [reply app/views/auth_sources/index.html.erb:9]: Unless we have added taxonomies to external usergroups there can be a lot of ldap auth sources and we want to show ldap card only one and not for all ldap authsources . If there is many ldap auth sources card will be created from first occurance (that prevents creating more ldap cards). If there is no ldap auth source temporary will be created because we need authsource object for card element. If it is confusing, I can add one more scope without ldaps and use that(for internal and external card)

File: app/views/auth_sources/_card.html.erb
Comment (ares) [reply app/views/auth_sources/_card.html.erb]: I'd put it under app/views/common

**Comment by dhlavac** [comment]
@ares Changes done

**Comment by dhlavac** [comment]
@ares Anything else what can I do on this PR ?

**Comment by mmoll** [comment]
@dhlavac please rebase :)
@ares any more comments?

**Comment by dhlavac** [comment]
@mmoll @ares rebased

**Comment by mmoll** [comment]
@dhlavac This needs a rebase :)

**Comment by dhlavac** [comment]
@mmoll I don't know if it is still planned to use this feature. @ares should I rebase that or close PR ?

**Comment by ezr-ondrej** [comment]
@dhlavac if you're not plannig to continue I was planning to take over. I do not know when I will get to it though.

**Comment by tbrisker** [comment]
closing in favor of #7206

**Comment by tbrisker** [comment]
Thanks for the effort here @dhlavac !

**Pull request closed without merging**
